rewrite-maven/src/main/java/org/openrewrite/maven/AddParentPom.java

```diff
@@ -0,0 +1,194 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.maven;
+
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+import org.openrewrite.*;
+import org.openrewrite.internal.StringUtils;
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.marker.SearchResult;
+import org.openrewrite.maven.table.MavenMetadataFailures;
+import org.openrewrite.maven.tree.MavenMetadata;
+import org.openrewrite.maven.tree.MavenRepository;
+import org.openrewrite.semver.Semver;
+import org.openrewrite.semver.VersionComparator;
+import org.openrewrite.xml.AddToTagVisitor;
+import org.openrewrite.xml.tree.Xml;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Value
+@EqualsAndHashCode(callSuper = false)
+public class AddParentPom extends Recipe {
+    transient MavenMetadataFailures metadataFailures = new MavenMetadataFailures(this);
+
+    @Option(displayName = "Group ID",
+            description = "The group ID of the maven parent pom to be adopted.",
+            example = "org.springframework.boot")
+    String groupId;
+
+    @Option(displayName = "Artifact ID",
+            description = "The artifact ID of the maven parent pom to be adopted.",
+            example = "spring-boot-starter-parent")
+    String artifactId;
+
+    @Option(displayName = "Version",
+            description = "An exact version number or node-style semver selector used to select the version number.",
+            example = "29.X")
+    String version;
+
+    @Option(displayName = "Relative path",
+            description = "New relative path attribute for parent lookup.",
+            example = "../pom.xml")
+    @Nullable
+    String relativePath;
+
+    @Option(displayName = "Version pattern",
+            description = "Allows version selection to be extended beyond the original Node Semver semantics. So for example," +
+                          "Setting 'version' to \"25-29\" can be paired with a metadata pattern of \"-jre\" to select Guava 29.0-jre",
+            example = "-jre",
+            required = false)
+    @Nullable
+    String versionPattern;
+
+    @Option(displayName = "File pattern",
+            description = "A glob expression that can be used to constrain which directories or source files should be searched. " +
+                          "Multiple patterns may be specified, separated by a semicolon `;`. " +
+                          "If multiple patterns are supplied any of the patterns matching will be interpreted as a match. " +
+                          "When not set, all source files are searched. ",
+            required = false,
+            example = "**/*-parent/grpc-*/pom.xml")
+    @Nullable
+    String filePattern;
```

Comment:
Typically we [use Preconditions](https://docs.openrewrite.org/reference/yaml-format-reference#preconditions) when we want to limit a recipe execution to a particular file. In particular it seems like a precondition of `org.openrewrite.FindSourceFiles` could be a replacement for this explicit option. Any arguments against dropping this option and nudging folks towards preconditions instead? That would be better for consistency, and ensures there's not proliferation of common options implemented in multiple repositories.

Reply:
OK, I'll remove it.

I'm new to OpenRewrite and didn't know about Preconditions. I added the `filePattern` field based on `AddPlugin` recipe.